src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowLoopSpec.groovy

```diff
@@ -0,0 +1,618 @@
+        northboundV2.createFlowLoop(flow.flowId, new FlowLoopPayload(switchPair.src.dpId))
+
+        when: "Try to create flowLoop on the dst switch"
+        northboundV2.createFlowLoop(flow.flowId, new FlowLoopPayload(switchPair.dst.dpId))
+
+        then: "FlowLoop is not created on the dst switch"
+        def exc2 = thrown(HttpClientErrorException)
+        exc2.statusCode == HttpStatus.BAD_REQUEST
+        with(exc2.responseBodyAsString.to(MessageError)) {
+            errorMessage == "Could not update flow"
+            errorDescription == "Can't change loop switch"
+        }
+
+        cleanup:
+        flow && flowHelperV2.deleteFlow(flow.flowId)
+    }
+
+    @Tidy
+    @Tags(LOW_PRIORITY)
+    def "Unable to create flowLoop on a transit switch"() {
+        given: "An active multi switch flow with transit switch"
+        def switchPair = topologyHelper.getNotNeighboringSwitchPair()
+        def flow = flowHelperV2.randomFlow(switchPair)
+        flowHelperV2.addFlow(flow)
+        def transitSwId = PathHelper.convert(northbound.getFlowPath(flow.flowId))[1].switchId
+
+        when: "Try to create flowLoop on the transit switch"
+        northboundV2.createFlowLoop(flow.flowId, new FlowLoopPayload(transitSwId))
+
+        then: "Human readable error is returned" //system can't update the flow when it is down
+        def exc = thrown(HttpClientErrorException)
+        exc.statusCode == HttpStatus.BAD_REQUEST
+        with(exc.responseBodyAsString.to(MessageError)) {
+            errorMessage == "Could not update flow"
+            errorDescription == "Loop switch is not terminating in flow path"
+        }
+
+        then: "FlowLoop is not created"
+        with(northbound.getFlow(flow.flowId)) {
+            !it.looped
+        }
+
+        and: "FlowLoop rules are not created on the dst switch"
+        Wrappers.wait(RULES_INSTALLATION_TIME) {
+            assert northbound.getSwitchRules(transitSwId).flowEntries.findAll {
+                new Cookie(it.cookie).getType() == CookieType.FLOW_LOOP
+            }.empty
+        }
+
+        cleanup:
+        flow && flowHelperV2.deleteFlow(flow.flowId)
+    }
+
+    def "getSwPairConnectedToTraffGenForSimpleFlow"(List<SwitchId> switchIds) {
+        getTopologyHelper().getAllNeighboringSwitchPairs().find {
+            [it.dst, it.src].every { it.dpId in switchIds }
+        }
+    }
+
+    def "getSwPairConnectedToTraffGenForProtectedFlow"(List<SwitchId> switchIds) {
+        getTopologyHelper().getAllNeighboringSwitchPairs().find {
+            [it.dst, it.src].every { it.dpId in switchIds } && it.paths.unique(false) {
+                a, b -> a.intersect(b) == [] ? 1 : 0
+            }.size() >= 2
+        }
+    }
+
+    def "getSwPairConnectedToTraffGenForVxlanFlow"(List<SwitchId> switchIds) {
+        getTopologyHelper().getAllNeighboringSwitchPairs().find {
+            [it.dst, it.src].every { it.dpId in switchIds && it.noviflow && !it.wb5164 }
```

Comment:
not an accurate way to pick a vxlan-enabled switch. Use switch features instead of manufacturer. Or even supported transit encapsulation, since it can support vxlan, but vxlan can be turned off